### DISCUSSION TITLE
License concerns

### DISCUSSION THREAD

**Discussion by hecksadecimal** [community/community#6638] [Copilot Conversations] [answered]
Given the nature of machine learning, I'm a little concerned about the use of copilot in any of my projects. There are cases where it can generate code that matches verbatim with that from another repository who's license may not be compatible with my own. The legality of it is questionable and it makes me hesitate to use it in any serious context. 

I could only find myself comfortable with it if there was an assurance that everything it was fine tuned on was MIT licensed. I'm not sure what else could be done to help set this worry to rest, but I felt like opening up a discussion about the implications.

**Comment by orph** [comment, marked as answer]
Hi there, there's a FAQ item which outlines our research into how rarely verbatim output occurs and the work we're doing to remove them altogether: https://copilot.github.com/#faq-does-github-copilot-recite-code-from-the-training-set

**Reply by mbasaglia** [reply]
how can you guarantee that verbatim code is not produced?
or even code that is substantially similar in the eye of copyright law?

**Reply by orph** [reply]
@mbasaglia great question! We've built a fast source code index so that Copilot can search for a proposed solution before displaying to users, and we've consulted with copyright specialists about details of how that system should work.

**Reply by CyFeng16** [reply]
> how can you guarantee that verbatim code is not produced? or even code that is substantially similar in the eye of copyright law?

They can't. Maybe you should ask how possible it will leak :)

**Comment by georgalis** [comment]
I share these concerns. We have bsd, gnu, libgnu, and a ton of other license in github open source code. There is even proprietary copyrighted code, which is open source, with unfettered access but for licensed use only.

I'm enthusiastic for the opportunity to try copilot, but I'm genuinely concerned the training may pick up and recommend novel methods which may be inappropriate (license wise) for my code base. 

I would love to see an ability to select specific licenses, or other parameters such as trusted authors, for the training used to derive the suggestions I receive. I expect this selection option could significantly impact the style and quality of code suggestions as well. User selectable training sets could provide a substantial win-win, by training with desired signal vs training from global signal combined with global noise.

**Comment by tupui** [comment]
It would be fairly easy for GitHub to just train on code which is explicitly marked as MIT/BSD. And as usual, training data being the most important thing when building a model, picking up manually the repo would be better to ensure that it's indeed the correct license, the code is good, etc.

As it is, this product is disrespectful of everyone. Dot. Cool that you had some talks with lawyers and whatever, I would be really curious to read some actual reports. I bet the discussion ended up as a tradeoff: money we would lose if we go to court vs money we are going to make... In any case there is something called ethic and sadly tech giants don't care that much about it.
